lib/Extractor/KLEEBuilder.cpp

```diff
@@ -188,11 +188,12 @@ ref<Expr> ExprBuilder::subnuwUB(Inst *I) {
 
 ref<Expr> ExprBuilder::mulnswUB(Inst *I) {
    const std::vector<Inst *> &Ops = I->orderedOps();
-   ref<Expr> L = SExtExpr::create(get(Ops[0]), 2*I->Width);
-   ref<Expr> R = SExtExpr::create(get(Ops[1]), 2*I->Width);
+   Expr::Width Width = get(Ops[0])->getWidth();
```

Comment:
Yes, the test case is test/Solver/overflow3.ll which is written to test
signed multiplication with overflow. In the implementation of SMULO
instruction which basically gives 1-bit overflow result, indicating whether
it overflowed or not.
  case Inst::SMulO:
    return XorExpr::create(mulnswUB(I), klee::ConstantExpr::create(1,
Expr::Bool));

I debugged and found I->Width gives me value 1 and expression width using
getWidth() gives me correct result of 32.


On Thu, Feb 26, 2015 at 11:01 AM, Raimondas Sasnauskas <
notifications@github.com> wrote:

> In lib/Extractor/KLEEBuilder.cpp
> <https://github.com/google/souper/pull/179#discussion_r25448830>:
>
> > @@ -188,11 +188,12 @@ ref<Expr> ExprBuilder::subnuwUB(Inst *I) {
> >
> >  ref<Expr> ExprBuilder::mulnswUB(Inst *I) {
> >     const std::vector<Inst *> &Ops = I->orderedOps();
> > -   ref<Expr> L = SExtExpr::create(get(Ops[0]), 2*I->Width);
> > -   ref<Expr> R = SExtExpr::create(get(Ops[1]), 2*I->Width);
> > +   Expr::Width Width = get(Ops[0])->getWidth();
>
> Can you please point me to that test case? The result of a mul inst is
> always the same width as the operands.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/souper/pull/179/files#r25448830>.
>



-- 
Thanks and Regards,
Jubi Taneja